Add community expectations about conduct and reviewing.

brendandburns opened the pull request:
@sarahnovotny @smarterclayton @philips @thockin @davidopp @jbeda 

As discussed in the previous community meeting here's a description of the expectations around community, I expanded it a little bit beyond code review because it seemed like a good place to put it
down in one location.

Please let me know what I got wrong (or right ;)

Thanks
--brendan

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32096)

<!-- Reviewable:end -->

brendandburns committed:
Add community expectations about conduct and reviewing.

thockin commented on community-expectations.md:
```diff
@@ -0,0 +1,85 @@
+Kubernetes is a community project and, as such, it is wholly dependent on
+its community to provide a productive, friendly and collaborative environment.
+
+The first and foremost goal of the Kubernetes community to develop orchestration
```
This is one heck of a sentence. The first part is missing a verb - probably 'is' ?

I would break and start a new sentence after "distributed systems".  "...distributed systems.  A second, equally important goal..."

misspelled orchestration

brendandburns replied:
done.

thockin commented on community-expectations.md:
```diff
@@ -0,0 +1,85 @@
+
+Any *active contributor* to the Kubernetes project is expected to be available
+to review code in areas where that contributor has expertise.  Active
+contributors or considered to be anyone who meets any of the following criteria:
```
s/or/are/

brendandburns replied:
done.

thockin commented on community-expectations.md:
```diff
@@ -0,0 +1,85 @@
+Any *active contributor* to the Kubernetes project is expected to be available
+to review code in areas where that contributor has expertise.  Active
+contributors or considered to be anyone who meets any of the following criteria:
+   * Sent more than two pull requests (PRs) in the previous one month, or more
```
I don't really like quantifying like this.  I am also worried about declaring "expectations" of people.  It feels condescending, somehow. I totally ACK the need to get more reviews, but the bar you have set is pretty low to have expectations...

davidopp replied:
+1, these numbers seem really arbitrary and could have the unintended consequence of causing people to hold back activity to avoid reaching it, if they really want to avoid reviewing.

I'd rather see expectation to do code reviews built into a more holsitic contributors ladder.

smarterclayton replied:
IF we're going to set exact numbers, we should double the margin in all directions.  Usually "a bar" is all that is required - if you said "1 PR in the last two months".

After the list, please describe the "why" of the list:

These criteria are intended to help us automate processes that deal with the large number of contributors to the project and helps us manage the project.  They are not intended to penalize members of the community with occasional contributions or who contribute in other spaces.

I.e. a bit more warm fuzzy.

thockin replied:
Agree with Clayton, that bar is too low.

Also, are we committing to tracking this?  Who is going to write the bot that adds and removes people from the reviewer pool?

I'm really uncomfortable with the numbers.  I am worried that, as soon as we start assigning reviews, this could be chilling.  Especially if the bar is low.  Someone who fixes typos, adds comments, clarifies errors, etc could EASILY hit 2 or 4 PRs in a month.  Someone who does some testing for their deployment could easily file 3 or more issues.  Commenting on PRs is cheap - do we want to discourage that?

This still feels like negative reinforcement to me, and we all know that doesn't work long term.  "If you want to play you have to review".  Can't we twist it around and just get the ladder defined - if you do X you are a "star reviewer"r.  If you do Y (> X), you are a "maintainer".  If you do Z (> Y) you are an "owner".  Positive reinforcement.

jbeda replied:
Another ideas -- some sort of "token" or "credit" to get your PR moved up in the queue based on your contribution to the community.  If you are both submitting and reviewing lots of PRs (not just code but other community work) then you get priority.

However, it is hard to do this without thinking about building automated systems to track it and there will always be room to game the system (small PRs for the sake of a PR, splitting a PR into multiple for credit).  Not sure how far down the path of quantification we want to go.

brendandburns replied:
I generally like the idea of special "silver bullets" or some sort of reward for positive reinforcement.

thockin commented on community-expectations.md:
```diff
@@ -0,0 +1,85 @@
+Reviewers are expected to respond in a timely fashion to PRs that are assigned
+to them.  Reviewers are expected to respond to an *active* PRs with reasonable
+latency, and if reviewers fail to respond, those PRs may be assigned to other
+reviewers.  If reviewers repeatedly fail to respond, then their own PRs (if
```
Again this just "feels" bad.  It feels petty.  I know we need something to motivate a broader base, but codifying it this way feels wrong.

davidopp replied:
+1, and what does "will also suffer increased latency" mean? Like we will intentionally delay reviews for people who aren't reviewing? How would you enforce that?

smarterclayton replied:
Agree we need to reword this in a slightly different way.  We should probably say that we want to properly incentivize reviewers and reward them, but that the mechanism for encouraging that is still under discussion.

davidopp commented:
Can this be combined this with one of the existing files in docs/devel? We have so many similar documents there... collab.md seems like it might be a good place for this.

davidopp commented:
(By similar I mean in the spirit of expectations for contributing to the project.)

eparis commented on community-expectations.md:
```diff
@@ -0,0 +1,85 @@
+The first and foremost goal of the Kubernetes community to develop orchestration
+technology that radically simplifies the process of creating reliable
+distributed systems, however a second, equally important goal is the creation
+of a community that fosters easy, agile development of such an orcheastration
```
s/orcheastration/orchestration/

jbeda replied:
This is a bit of a run-on sentence.  "... systems.  A second, equally important, goal is the..."

smarterclayton commented on community-expectations.md:
```diff
@@ -0,0 +1,85 @@
+## Community Expectations
+
+Kubernetes is a community project and, as such, it is wholly dependent on
```
remove `, as such, it `

brendandburns replied:
done.

smarterclayton commented on community-expectations.md:
```diff
@@ -0,0 +1,85 @@
+of a community that fosters easy, agile development of such an orcheastration
+systems.
+
+This document describes the community expectations for
```
I think it's critically important that the first sentence here needs to be a royal we - i.e. "We the community therefore describe our expectations for our members."   It can't be dispassionate outsider.

jbeda replied:
Agree -- perhaps invite community members to open PRs to change this document as the expectations in our community evolve.

brendandburns replied:
done.

krmayankk commented:
great start with good intentions.  its interesting how we will enforce any of the things mentioned above. We will always have some people highly motivated to do both reviews and PR, some who will be just for one or the other. There will be some people who just need one feature, and they will follow up until that feature they want is merged or issue fixed and in the process will develop deep expertise around that feature and once that is done, they have other critical issues on their plate from the organization they represent, that they will move on. They will be active contributors during that period.

roberthbailey commented on community-expectations.md:
```diff
@@ -0,0 +1,85 @@
+
+#### Expectations of reviewers: Review latency
+
+Reviewers are expected to respond in a timely fashion to PRs that are assigned
```
Any chance we can have the reviewer assignment honor an OOO config somewhere? I've probably been ignoring PRs for the last month while on leave.....

brendandburns replied:
Yes, this has been requested elsewhere and will definitely be part of the system.

jbeda replied:
I'd love to find a way to help reviewers know when they are falling behind. I've been failing to deal with the deluge of github mail and have been missing issues that have been assigned to me or issues that I've been mentioned in.

brendandburns commented:
@thockin @davidopp I think that is super important that it is clear that these are expectations, it's not optional.

And in particular, we need hard numbers because Brandon is going to run a script and we're going to automatically opt-in people to reviewing who meet this criteria.

This is a big part of why I brought this up in community meeting, we need to decide (as a community) if we are going to require people who are active in the community to do code review.  It seems to me that we can either maintain the status quo where fewer people review than contribute and therefore review latency and load is high, or we can go to something like this, where active participation in the community means active participation in all parts of the community.

By all means we can try an "all carrots" approach whereby we don't penalize people who don't do reviews but we do reward people who do a lot of reviews (not exactly sure what that looks like, but I'm sure it can be done)

But I think we are going to have to quantify what "active" means, if only so that we can write the code that opts people in to being reviewers.

Paging @sarahnovotny for her thoughts on this.

philips commented:
For everyone's context this conversation was started when we discussed merging a list of reviewers into the OWNERS files (details on https://github.com/kubernetes/kubernetes/pull/31567). The motivation for introducing reviewers is to increase velocity by sharding the toil of code review across more people.

However, when we put this PR forward to add reviewers two things happened:

1. People wanted to move to an opt-in vs. opt-out model for reviewers
2. People debated the improvement that the reviewers system will deliver without a carrot

In the Kubernetes community I have very often heard the refrain "if people review more velocity will increase and your stuff will get merged faster". Now, I think there is some glimpse of truth to that but I think there are blocks to making that true:

1. Quality reviews by a small group that the owner/assignee trusts
2. Documented community expectation about the job of reviewers
3. A carrot or stick to encourage reviewer-ship an activity folks do

Today, I feel like 1 & 2 are sufficient to get things going. And I think we should hold off on 3 until we see how well or poorly 1&2 work.

Given all that, I think we should:

- Merge this PR minus carrot/sticks so the people doing the work to automate the collection of reviewers in https://github.com/kubernetes/kubernetes/pull/31567 and the people being volunteered to review know the why/how.
- Hold off on the carrot/stick conversation until we see how this first round of getting people to do reviews go. If we later feel we need a stick/carrot we can add one.

brendandburns commented:
@philips thanks for your thoughts.

I'm fine w/ merging this minus the carrots/sticks and see how we do.

@thockin @davidopp what do you think?  If you agree, I'll revise this PR to remove carrots/sticks but retain the numbers to quantify what "active" means.

sarahnovotny commented:
I like the definitions of what active is.  If we have metrics, we can follow with the more subjective lense "are these people contributing in different but important ways that are ok or better".  

I see this as doing 2 things 

1. encouraging more reviews by *everyone* to grow the reviewer base, and decrease the load on the maintainers. ideally, though probably eventually, increasing the velocity of the review/merge process

but, more importantly

2. to showcase the skills of the new/different contributors as a means to establish explicitly referenceable trust and eventually grow the maintainer ranks.

@davidopp has a good point that we need to consolidate and rationalize the many places we're collecting what contributing means type of docs... I have talked to both the Intel and CoreOS lots (via @ConnorDoyle and @philips) about having their next newbies help with that as they join the project and learn the cultural expectations.

brendandburns commented:
Ping to @thockin and @davidopp

thockin commented:
There are a lot of comments still open on the doc, but my strongest feeling has been restated.  I'll defer to @sarahnovotny . If you say we need this, I'll follow your lead, but all my "training" as a parent is kicking in saying this is structured backwards.

ConnorDoyle commented:
Just have 2c to drop into the conversation: all reviews not equal. e.g. "carrot" motivated reviews are likely to be of much higher quality than "stick" motivated reviews. So although either method might yield the same number the qualitative part, which the bot cannot measure easily, could vary a lot.

philips commented:
@thockin I think @brendandburns was asking if you think it is OK to set expectations while dropping the carrot and stick. 

Can you explain a bit more about "structured backwards"?

thockin commented:
"I expect you to do X" implies "or else ...", even if we don't say it.
That is less effective than "if you do X, good-thing Y".  Positive
reinforcement is a stronger motivator than negative reinforcement, and it
lasts longer.  Building up an identity around "I'm a {kontributor,
kollaborator, kommitter} is stronger yet.

Obviously, people are complex animals, but it's well understood that
intrinsic motivators and positive reinforcement are strong.  I don't want
people to feel the implied "or else".

On Thu, Sep 15, 2016 at 6:15 AM, Brandon Philips <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> I think @brendandburns
> <https://github.com/brendandburns> was asking if you think it is OK to
> set expectations while dropping the carrot and stick.
>
> Can you explain a bit more about "structured backwards"?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32096#issuecomment-247324396>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVJVkeQ-s7qpUB4wUBEnIcYa2j0bJks5qqUTkgaJpZM4J1br8>
> .
>

philips commented:
On Thu, Sep 15, 2016 at 9:38 AM Tim Hockin <notifications@github.com> wrote:

> "I expect you to do X" implies "or else ...", even if we don't say it.
> That is less effective than "if you do X, good-thing Y". Positive
> reinforcement is a stronger motivator than negative reinforcement, and it
> lasts longer. Building up an identity around "I'm a {kontributor,
> kollaborator, kommitter} is stronger yet.
>
> Obviously, people are complex animals, but it's well understood that
> intrinsic motivators and positive reinforcement are strong. I don't want
> people to feel the implied "or else".
>

Got it. Can we simply say something loose like: "If you do code reviews it
will help ensure Kubernetes continues to make solid headline worthy
releases. A track record of code reviews is also an important measure used
when considering adding people as subsystem approvers."

I really just want to try and get this new reviewer system in place before
trying to get an explicit carrot into place. We don't even know if a carrot
is necessary but we do know there are smart engaged folks who's eyes could
be pointed at relevant PRs to review.

Brandon

thockin commented:
If smart, engaged, willing folks are waiting for a doc to tell then they
are expected to look at code reviews, I don't know what to say.  As
written, this reads to me as a "chore sheet" - something that we need to
point the less-engaged, less-willing people towards.

@sarahnovotny what is between where-we-are and a proper contributor ladder
definition?  If we're going to quantify, let's quantify in a way that
promotes people for good work rather than assigns them work in order to be
promoted.

On Thu, Sep 15, 2016 at 1:27 PM, Brandon Philips <notifications@github.com>
wrote:

> On Thu, Sep 15, 2016 at 9:38 AM Tim Hockin <notifications@github.com>
> wrote:
>
> > "I expect you to do X" implies "or else ...", even if we don't say it.
> > That is less effective than "if you do X, good-thing Y". Positive
> > reinforcement is a stronger motivator than negative reinforcement, and it
> > lasts longer. Building up an identity around "I'm a {kontributor,
> > kollaborator, kommitter} is stronger yet.
> >
> > Obviously, people are complex animals, but it's well understood that
> > intrinsic motivators and positive reinforcement are strong. I don't want
> > people to feel the implied "or else".
> >
>
> Got it. Can we simply say something loose like: "If you do code reviews it
> will help ensure Kubernetes continues to make solid headline worthy
> releases. A track record of code reviews is also an important measure used
> when considering adding people as subsystem approvers."
>
> I really just want to try and get this new reviewer system in place before
> trying to get an explicit carrot into place. We don't even know if a carrot
> is necessary but we do know there are smart engaged folks who's eyes could
> be pointed at relevant PRs to review.
>
> Brandon
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32096#issuecomment-247444541>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVJXcAlTSpYb6f93IuBeH2k3J-S8Lks5qqao0gaJpZM4J1br8>
> .
>

brendandburns commented:
@thockin, at the end of the day I'm looking for two things:
   * A page I can point people at when they say "Why did you opt me into this reviewers thingy"
   * A set of numbers that we can plug into automation to decide who is the reviewer (@phillips et al have already built this automation)

thockin commented:
If we're opting people into reviews, it's going to end in tears.

On Thu, Sep 15, 2016 at 8:01 PM, Brendan Burns <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin>, at the end of the day I'm looking
> for two things:
>
>    - A page I can point people at when they say "Why did you opt me into
>    this reviewers thingy"
>    - A set of numbers that we can plug into automation to decide who is
>    the reviewer (@phillips <https://github.com/phillips> et al have
>    already built this automation)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32096#issuecomment-247507662>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVE6kR4ABduYAituRsAoCl663ac1bks5qqgajgaJpZM4J1br8>
> .
>

eparis commented:
opting people in has 2 aspects.

1. it will make people who may not feel entitled to do reviews feel a sense of ownership and appropriateness that they may not grant themselves. Not everyone has as high of an opinion of themselves as I do and not everyone feels they should be reviewing others and telling others what they did wrong. Opting people in grants them permission and responsibility.
2. opting people in also means that we get me people like me who are terrible reviewers. I'd love to see the accept/reject review thing which was discussed at one point.

Overall I think the tradeoff between empowering/encouraging our community to be leaders and the potential lack of response will be better if we both opt people in and then opt people out who aren't doing such a great job following through. We can't force reviews on people and expect it to always be a positive, but we definitely need a way to prod people up the kubernetes leadership ladder.

I fall on the side of 'opt people in' but hope we track how it goes and make a commitment to iterate.

philips commented:
Right, @brendandburns we have to have something to write in the email for the potential reviewers and some metric to finalize the selection process. If we get it wrong it is OK to iterate from here; most of the targeted people will already be used to us and probably just opt-in without reading :-P

Brendan, can you drop the carrot/stick and fix up the nits and push?

philips commented:
@thockin no one is being surprised opted-in. We will email them (see https://github.com/kubernetes/kubernetes/pull/31752) and based on the current metrics I think about ~100 people will be emailed.

thockin commented:
100 new reviewers seems implausible to me - it suggests the metrics are too
low.

I don't want to block progress, I understand exactly how much we need more
first-line reviewers.  I am just trying wear the hat of a casual
contributor, and being emailed "we expect you to do code reviews" would be
weird, to me.  It feels like it is 180 degrees from what I would expect - I
should be offering to do reviews (or just doing them), rather than being
told to.  What we are missing is some extrinsic motivators.

I will STFU on this, and let it play out if the smart people herein all
feel I am wrong.

On Sep 16, 2016 1:24 PM, "Brandon Philips" <notifications@github.com> wrote:

> @thockin <https://github.com/thockin> no one is being surprised opted-in.
> We will email them (see #31752
> <https://github.com/kubernetes/kubernetes/pull/31752>) and based on the
> current metrics I think about ~100 people will be emailed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32096#issuecomment-247699998>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVKufAyCV1S44CoOkOYEhbButqUjWks5qqvrwgaJpZM4J1br8>
> .
>

davidopp commented:
I didn't read the whole thread yet but I agree with Tim insofar as opting people in to do reviews seems like it will have a number of negative consequences. Among them, it seems like it will promote quantity over quality, leaving it to the "owners" to do the real reviews.

There must be less draconian ways to prevent "tragedy of the commons." For example, since the real bottleneck is likely to be owners/committers, maybe we could say your PRs will get expedited attention from owner/committers if you are a solid reviewer. This is just an example; I'm sure there are better ideas.

philips commented:
On Sat, Sep 17, 2016 at 11:44 AM Tim Hockin <notifications@github.com>
wrote:

> 100 new reviewers seems implausible to me - it suggests the metrics are too
> low.
>

Metrics are too low? I don't understand.


> I don't want to block progress, I understand exactly how much we need more
> first-line reviewers. I am just trying wear the hat of a casual
> contributor, and being emailed "we expect you to do code reviews" would be
> weird, to me. It feels like it is 180 degrees from what I would expect - I
> should be offering to do reviews (or just doing them), rather than being
> told to. What we are missing is some extrinsic motivators.
>

I feel this PR, with the removal of the carrot/stick at the end, is a good
starting spot to set expectations, outline the reviewer selection criteria,
and document the reviewer process. @brendandburns can you please remove
this carrot/stick and force push?

The email that needs to go out essentially looks like this:

- We need help doing reviews.
- We know in the past finding work to review has been difficult so we built
automation.
- You were selected, based on previous work, to help review these
subsystems.
- If you opt-in a bot is going to ping you to do reviews on these subsystems

Can @thockin or @davidopp come up with a proposal document that is better
at expressing these bullet points?

Having more people empowered with an explicit task list of things they
should be doing in the community is useful, I think. We are a large
organization now and large organizations generally need to move from
implicit to explicit expectations.

thockin commented:
On Sep 17, 2016 2:40 PM, "Brandon Philips" <notifications@github.com> wrote:
>
> On Sat, Sep 17, 2016 at 11:44 AM Tim Hockin <notifications@github.com>
> wrote:
>
> > 100 new reviewers seems implausible to me - it suggests the metrics are
too
> > low.
> >
>
> Metrics are too low? I don't understand.

The numbers used.  The net is too wide and catches way more people than is
reasonable.  I think the initial cut of this should catch a dozen-ish
people who would make good reviewers.

>
> > I don't want to block progress, I understand exactly how much we need
more
> > first-line reviewers. I am just trying wear the hat of a casual
> > contributor, and being emailed "we expect you to do code reviews" would
be
> > weird, to me. It feels like it is 180 degrees from what I would expect
- I
> > should be offering to do reviews (or just doing them), rather than being
> > told to. What we are missing is some extrinsic motivators.
> >
>
> I feel this PR, with the removal of the carrot/stick at the end, is a good
> starting spot to set expectations, outline the reviewer selection
criteria,
> and document the reviewer process. @brendandburns can you please remove
> this carrot/stick and force push?
>
> The email that needs to go out essentially looks like this:
>
> - We need help doing reviews.
> - We know in the past finding work to review has been difficult so we
built
> automation.
> - You were selected, based on previous work, to help review these
> subsystems.
> - If you opt-in a bot is going to ping you to do reviews on these
subsystems

I am ok with this, but not to 100 people.

> Can @thockin or @davidopp come up with a proposal document that is better
> at expressing these bullet points?
>
> Having more people empowered with an explicit task list of things they
> should be doing in the community is useful, I think. We are a large
> organization now and large organizations generally need to move from
> implicit to explicit expectations.

ACK people are much more complex than software.

brendandburns commented:
I'll update this PR today to remove carrot/stick.

@thockin do you want there to be fewer than 100 active reviewers?  I think if we go < 100 we're going to stick with the status quo (~50 Google, ~30 Red Hat, ~20 community)

--brendan

sarahnovotny commented:
I've generally considered more reviewers as a way to grow trust.  not
immediately presuming that each new review/reviewer is useful, but over
time with secondary (more experienced) reviewers and approvers commenting,
the newly added reviewers learn and the existing approvers see who has the
chops to just get LGTM'd based on prior track record of (extensive and
good) reviews.

sarah

On Mon, Sep 26, 2016 at 8:56 AM Brendan Burns <notifications@github.com>
wrote:

> I'll update this PR today to remove carrot/stick.
>
> @thockin <https://github.com/thockin> do you want there to be fewer than
> 100 active reviewers? I think if we go < 100 we're going to stick with the
> status quo (~50 Google, ~30 Red Hat, ~20 community)
>
> --brendan
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32096#issuecomment-249612478>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHxirdcRAtRsl2sm_zAvNrlp5JRLNgfks5qt-s0gaJpZM4J1br8>
> .
>

jbeda commented:
I'm a bit late to the party here (sorry!) but how about this.  I apologize if this is well tread ground.

* We define Kubernetes ranks.  
* Those ranks are both based on actions *and* responsibilities.
* People need to opt in to a rank.
* Ranks come with privileges.

Above a certain rank you'll have an expectation to do so many reviews and other work.  You'll also get certain privileges -- the ability to prioritize your own PRs, swag, recognition, fame, guaranteed invites to dev events, etc.

We can make a rank something that folks can put on their resume so that they can demonstrate they are an active real member of the community.

thockin commented:
That's the contributor ladder that Sarah is working on.  There's some
urgency (??) to get this step in first...

On Mon, Sep 26, 2016 at 9:35 AM, Joe Beda <notifications@github.com> wrote:

> I'm a bit late to the party here (sorry!) but how about this. I apologize
> if this is well tread ground.
>
>    - We define Kubernetes ranks.
>    - Those ranks are both based on actions *and* responsibilities.
>    - People need to opt in to a rank.
>    - Ranks come with privileges.
>
> Above a certain rank you'll have an expectation to do so many reviews and
> other work. You'll also get certain privileges -- the ability to prioritize
> your own PRs, swag, recognition, fame, guaranteed invites to dev events,
> etc.
>
> We can make a rank something that folks can put on their resume so that
> they can demonstrate they are an active real member of the community.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32096#issuecomment-249623576>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVF0ElcdWnkFb0IyA7KAOvhiNQxSrks5qt_RPgaJpZM4J1br8>
> .
>

thockin commented:
My point was that if this catches 100 NEW reviewers, we're doing it wrong.
If that 100 is all the people already doing reviews plus 20 community
folks, OK.

On Mon, Sep 26, 2016 at 8:56 AM, Brendan Burns <notifications@github.com>
wrote:

> I'll update this PR today to remove carrot/stick.
>
> @thockin <https://github.com/thockin> do you want there to be fewer than
> 100 active reviewers? I think if we go < 100 we're going to stick with the
> status quo (~50 Google, ~30 Red Hat, ~20 community)
>
> --brendan
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32096#issuecomment-249612478>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVGuVo3RXRw5iCBvhjixgoQopVvFTks5qt-s0gaJpZM4J1br8>
> .
>

brendandburns commented:
@thockin agree, the original PR added ~150 more reviewers which triggered my desire to write this doc. ;)

thockin commented:
I guess my lingering concern is just the threshold tuning, then.  Do we
have numbers/graphs on the parameters?

On Mon, Sep 26, 2016 at 9:48 PM, Brendan Burns <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> agree, the original PR added ~150
> more reviewers which triggered my desire to write this doc. ;)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32096#issuecomment-249765481>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVPpqEAi3ZtcElZ6CipyaqC68FWueks5quKASgaJpZM4J1br8>
> .
>

brendandburns commented:
@thockin @davidopp @philips @jbeda @sarahnovotny @smarterclayton 

Comments addressed, carrots and sticks removed (I think)

Please take another look.

Thanks
--brendan

brendandburns commented:
pretty sure @philips can get us parameters.  Brandon, do you want to tune the parameters for adding ~20 additional reviewers who are not currently active?  (active reviewer defined as having been assigned a review once in the last month?)

Thanks!

bgrant0607 commented:
Sorry, I was OOO. If this is still open in 2 days, I'll try to get to it then.

bgrant0607 commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+latency, and if reviewers fail to respond, those PRs may be assigned to other
+reviewers.
+
+*Active* PRs are considered those which have a proper CLA (`cla:yes`) label
```
FYI, we're working on a PR workflow that defines states.
https://docs.google.com/document/d/194MGgJHlm4ovXtvzvkjzzBZG_v47d95nsjuuaUe1Log/edit?ts=57929ed1#heading=h.h8u91v47whwa

davidopp replied:
@mml wrote a cool script for this, and I have seen at least one other, though I can't remember where.

bgrant0607 commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+## Community Expectations
```
Could this doc please go into docs/devel?

brendandburns replied:
I'm going to wait until I get LGTM on the doc in order to preserve comment history, then I'll move it...

brendandburns replied:
moved.

thockin commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+Any *active contributor* to the Kubernetes project is expected to be available
+to review code in areas where that contributor has expertise.  Active
+contributors are considered to be anyone who meets any of the following criteria:
+   * Sent more than two pull requests (PRs) in the previous one month, or more
```
nit: you spelled "two" here and used numeric "3" below

thockin commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+Consequently, as a community we expect that all active participants in the
+community will also be active reviewers.
+
+Any *active contributor* to the Kubernetes project is expected to be available
```
"expected to be available" just sounds so harsh.  How about "We ask that *active contributors* to the project participate in the code review process in areas where that contributor has expertise."

Too wimpy?

davidopp replied:
I like Tim's rewording

bgrant0607 replied:
I like Tim's wording, also, because it sounds more active ("participate").

thockin commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+   than 20 PRs in the previous year.
+   * Filed more than 3 issues in the previous month, or more than 30 issues in
+   the previous year.
+   * Commented on more than 5 issues or pull requests in the previous month, or
```
This is the one I feel will cast too wide a net.  I don't want to discourage commenting...

davidopp replied:
I think these are all too specific. They're just begging for unintended consequences IMO. But I agree this one is the most problematic.

brendandburns replied:
Maybe I'll make this "Commented on more than 5 PRs?"  I think if you are actively commenting on PRs then you are by definition an active reviewer and should be considered for the pool.

Also there are several people in the project who do a lot of reviewing and commenting, but often don't actually write any PRs.

brendandburns replied:
removed issues, focused on PRs alone.

thockin reviewed (commented):
I'll go along with it.  Just some nits.

davidopp commented:
I'm still not thrilled that this is a separate doc, given that we already have a bazillion docs in docs/devel/ on "how to contribute to Kubernetes." But a bazillion plus one is not much worse than a bazillion, so I guess it doesn't matter.

davidopp commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+The first and foremost goal of the Kubernetes community to develop orchestration
+technology that radically simplifies the process of creating reliable
+distributed systems. However a second, equally important goal is the creation
+of a community that fosters easy, agile development of such an orchestration
```
s/an//

brendandburns replied:
done.

davidopp commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+   the previous year.
+   * Commented on more than 5 issues or pull requests in the previous month, or
+   more than 50 issues or pull requests in the previous year.
+   * Marked any PR as LGTM in the previous month.
```
I didn't understand this one. What does it mean for someone who is not a reviewer to mark something as LGTM?

bgrant0607 replied:
It's possible to mark PRs as LGTM without otherwise commenting.

brendandburns replied:
Also, if you LGTM one thing, that makes you active, even if that's your only comment on a PR in that month.

brendandburns commented:
I'll address comments today and push a (hopefully) final draft.

@davidopp if you think there is a better home, let me know and I'll move it there.

I personally think this is a more important document than most because it shapes the shape of the community, but that's my thought.

bgrant0607 commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+to review code in areas where that contributor has expertise.  Active
+contributors are considered to be anyone who meets any of the following criteria:
+   * Sent more than two pull requests (PRs) in the previous one month, or more
+   than 20 PRs in the previous year.
```
Rather than "previous year", how about something like "trailing 12 months", to resolve confusion about whether today this would mean in 2015 or Sept 2015 to Sept 2016.

brendandburns replied:
moved to "previous 12 months"

bgrant0607 commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+   * Commented on more than 5 issues or pull requests in the previous month, or
+   more than 50 issues or pull requests in the previous year.
+   * Marked any PR as LGTM in the previous month.
+   * Have *contributor* permissions in the Kubernetes github project.
```
What are "contributor" permissions? Write or admin access?

brendandburns replied:
ah, 'collaborator' is the github term: https://help.github.com/articles/permission-levels-for-a-user-account-repository/

switched to that.

bgrant0607 commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+
+Reviewers are expected to respond in a timely fashion to PRs that are assigned
+to them.  Reviewers are expected to respond to an *active* PRs with reasonable
+latency, and if reviewers fail to respond, those PRs may be assigned to other
```
I plan to automatically reassign such PRs and to track reviewer stats.

brendandburns replied:
Yeah, I think this is what we want in the limit.

bgrant0607 commented:
@k8s-bot test this issue: #33481

bgrant0607 commented:
I made some comments, but would like to merge this and iterate. If more changes are made, great. If not, I'll LGTM anyway later today.

brendandburns commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+of a community that fosters easy, agile development of such an orchestration
+systems.
+
+We the community therefore describe the expectations for
```
done.

brendandburns commented on community-expectations.md:
```diff
@@ -0,0 +1,83 @@
+
+### Code review
+
+As a community we believe in the value of code review for all contributions.
```
done.

brendandburns commented:
Ok, I addressed all of the comments, and squashed, this is ready for merge as far as I'm concerned.

thanks all for the feedback, please let me know what you'd like to see in further revisions.

bgrant0607 commented:
Verification checks failed. Please run hack/update-munge-docs.sh

bgrant0607 commented:
LGTM otherwise. Feel free to re-apply after you push fixes.

brendandburns commented:
update-munge-docs run, re-applied 'lgtm'

philips commented:
LGTM. I wish there was some way to provide a link to all PRs that are active for a reviewer but that will require more tooling. Ship it!
